Add standalone man page builder

mattclay opened the pull request:
##### SUMMARY

Add standalone man page builder.

Resolves https://github.com/ansible/ansible/issues/80368

##### ISSUE TYPE

Feature Pull Request

##### COMPONENT NAME

packaging/pep517_backend/

mattclay committed:
Add standalone man page builder

gotmax23 commented on packaging/pep517_backend/man.py:
```diff
@@ -0,0 +1,73 @@
+            _convert_rst_to_man(rst_doc_template, destination_path)
+
+
+def _generate_rst_in_templates(output_dir: pathlib.Path) -> t.Iterable[pathlib.Path]:
```
`typing.Iterable` is deprecated. You should use `collections.abc.Iterable[pathlib.Path]` here instead.

gotmax23 commented on packaging/pep517_backend/man.py:
```diff
@@ -0,0 +1,73 @@
+        *entry_points,
+    )
+
+    subprocess.run(tuple(map(str, generate_man_cmd)), check=True)
```
You don't need to convert Path objects into strings. `subprocess.run` accepts PathLike objects in addition to strings.

gotmax23 commented on packaging/pep517_backend/man.py:
```diff
@@ -0,0 +1,73 @@
+
+def _generate_rst_in_templates(output_dir: pathlib.Path) -> t.Iterable[pathlib.Path]:
+    """Create ``*.1.rst.in`` files out of CLI Python modules."""
+    entry_points = list(pathlib.Path(CHECKOUT_DIR / 'lib/ansible/cli').glob('*.py'))
```
`CHECKOUT_DIR / 'lib/ansible/cli'` already returns a Path.  Also, I'm sure what's the point in converting this to a list; the code only iterates over it once.

```suggestion
    entry_points = (CHECKOUT_DIR / 'lib/ansible/cli').glob('*.py')
```

gotmax23 reviewed (commented):
Thanks! This should work for us. I left a couple comments :).

mattclay committed:
Minor refactoring

webknjaz commented on packaging/pep517_backend/_backend.py:
```diff
@@ -122,15 +79,9 @@ def build_sdist(  # noqa: WPS210, WPS430
-                    version_number=version_number,
-                )
-                rst_in.unlink()
+            from .man import build_man_pages, DEFAULT_RELATIVE_OUTPUT_DIR
```
I'd rather not allow non-module-level imports.

webknjaz commented on packaging/pep517_backend/man.py:
```diff
@@ -0,0 +1,69 @@
+    build_man_pages(__version__, CHECKOUT_DIR / DEFAULT_RELATIVE_OUTPUT_DIR)
+
+
+def build_man_pages(version: str, output_dir: pathlib.Path) -> None:
```
`output_dir` could default to cwd and the outer call would not have to pass a concatenated dir, since it does a `chdir` anyway.

webknjaz commented on packaging/pep517_backend/man.py:
If this was to be accepted, it would probably have to live outside of the PEP 517 backend...

webknjaz reviewed (changes requested):
I've made a few tech comments below but I'd rather take a step back first, to check the motivation and consequences of going this way.

To summarize the comments/observations I've made elsewhere, in short:
1) We made effort in the past to get rid of downstream-specific packaging definitions/hacks
2) We made an effort to expose packaging interfaces that are standardized and favored by the larger Python community
3) What we implemented provides a way for the downstreams (and anyone, really) to create a correct source distribution (don't confuse with a source archive!)
4) We publish correct source distributions to PyPI
5) Most downstreams use those official source distributions and are able to extract what's necessary from them
6) All downstreams I checked are good with PEP 517 (even Debian, despite having a bad track record of packaging Python-related things in the past)
7) It is possible for Fedora to either use the official source distribution or replicate our build process that creates the distribution artifacts
8) We test our release automation, specifically, the PEP 517 bits
9) We do not test alternative invocations that might make use of it, and never did
10) We actively discouraged using Git archives as a source for installing Ansible in the past, at least; I understand it was broken'ish
11) We do not test installing from Git archives

So basically, we've made an effort to distantiate ourselves from the downstream decision-making while providing a community-accepted packaging interfaces.

My understanding is that Fedora made a conscious decision to disregard our earlier instructions and go the untested discouraged way. It appears to me that downstreams should not shifting the maintenance burden to the upstreams with regard to things that do not bring value to upstreams. It seems, this could currently only facilitate Fedora packagers but not others. Based on our previous efforts not to keep downstream-specific things in the upstream repository, as well as providing the standard packaging interface, it seems like maintaining two ways of doing the same thing is not sustainable.

Moreover, if we were to do this, we'd have to basically extend our testing of:
1) Installing from Git archives
2) Running Fedora-specific entrypoints for packaging

Based on above, I'm inclined to recommend that we reject the original proposal to maintain an extra entry point.

mattclay committed:
Separate man page builder from backend

mattclay commented:
I moved the man page generator out of the backend, so it's no longer split halfway between the `hacking` directory and the `packaging` directory. Now the backend just calls a subprocess (as the current implementation does). The only difference is that the subprocess handles all the man page generation, instead of just the templates.

mattclay committed:
Minor code cleanup

mattclay commented:
With this change, the man page generation is more consolidated, which is a good thing. It just happens to resolve https://github.com/ansible/ansible/issues/80368 as well. Even if we didn't want to resolve that issue, I think this change still makes sense. I think it also also addresses the concerns @webknjaz expressed.

gotmax23 commented on packaging/pep517_backend/_backend.py:
```diff
@@ -122,15 +81,14 @@ def build_sdist(  # noqa: WPS210, WPS430
-                rst_in.unlink()
+
+            build_man_pages_command = (
+                pathlib.Path(tmp_src_dir) / 'hacking/build-man-pages.py',
```
```suggestion
                tmp_src_dir / 'hacking/build-man-pages.py',
```

mattclay committed:
Use check_call like the original code did

webknjaz commented on packaging/pep517_backend/_backend.py:
```diff
@@ -32,12 +31,6 @@ def _chdir_cm(path: os.PathLike) -> t.Iterator[None]:
-with suppress(ImportError):
-    # NOTE: Only available for sdist builds that bundle manpages. Declared by
-    # NOTE: `get_requires_for_build_sdist()` when `--build-manpages` is passed.
-    from docutils.core import publish_file
```
Let's just put importing `build_man_pages` under suppress here.

gotmax23 replied:
Wouldn't it make more sense to avoid `suppress(ImportError)`? If someone e.g. builds the sdist with `python -m build --no-isolation` and docutils isn't available, they'll get a NameError when the code calls the missing function which would be quite confusing.

webknjaz replied:
@gotmax23 not at all. If someone chooses not to use pypa/build's built-in virtualenv provisioning, it does not mean they shouldn't provide them. `--no-isolation` only disables the throwaway virtualenv management, putting the responsibility of provisioning them onto the invoker. Besides, `--no-isolation` does not imply `--skip-dependency-check` IIRC, so pypa/build will still call the corresponding hooks to learn which deps are mandatory and check their presence, I think.
But even if you disable both, the code calling these imported bits is only invoked with `--config-setting=--build-manpages`, meaning that `NameError` will only happen under the feature flag — this is when the control flow would actually hit the lines referring to the undefined variables in runtime. Python does not (and cannot) evaluate the references until it reaches them. It'll hit function definitions on import but not their internal scope.

Similarly, if one forgets to install `setuptools`, the same thing is going to happen.

FWIW the pyproject macros should probably provide a way to extract deps needed for building sdists.
The current reusable implementation of calling hooks provides this via https://pyproject-hooks.rtfd.io/en/latest/pyproject_hooks.html#pyproject_hooks.BuildBackendHookCaller.get_requires_for_build_sdist

gotmax23 replied:
>  meaning that NameError will only happen under the feature flag

Correct.

> FWIW the pyproject macros should probably provide a way to extract deps needed for building sdists.

The macros call `pip wheel` directly and don't build sdists, so they don't need to generate those deps. If you have any feature/bug requests for the pyproject-rpm-macros project, you should [file a Bugzilla](https://bugzilla.redhat.com/enter_bug.cgi?product=Fedora&component=pyproject-rpm-macros).

webknjaz replied:
> The macros call `pip wheel` directly and don't build sdists, so they don't need to generate those deps.

I know, but they could do so to facilitate your use-case where you're reproducing some of the sdist build processes so you wouldn't have to hardcode the respective deps declaration.

webknjaz commented on MANIFEST.in:
```diff
@@ -45,6 +45,7 @@ include MANIFEST.in
 include changelogs/changelog.yaml
 recursive-include hacking/build_library *.py
+include hacking/build-man-pages.py
 include hacking/build-ansible.py
```
Side note: why do we need to include these things in sdist? Do we expect somebody to untar it and use them directly? Why?

mattclay replied:
We don't need to. We could just add it to the ignore list instead:

 https://github.com/ansible/ansible/blob/d7e7c2d87260bc3abe0dbb2725b34d4469860b3d/test/sanity/code-smell/package-data.py#L44

gotmax23 replied:
I'd rather include it. People may want to delete the manpages and regenerate them.

mattclay replied:
I generally lean towards including everything in the sdist unless there's some reason why it doesn't make sense, such as our CI configuration.

bcoca replied:
the whole point of hacking/ was to put there things no one would want/use in the dist, stuff we only use internally 

should we move these things out of hacking/?

webknjaz replied:
@bcoca yes, that's what we're trying to do.

webknjaz commented on hacking/build-man-pages.py:
I'd split this into 2 parts:
1) Have entrypoint/hacking CLI script under `hacking/` — this bit is not important for PEP 517
2) Keep the manpage generation logic module with a single importable function as its API under the in-tree back-end since its the component where we use it and it's kinda its part

webknjaz commented on hacking/build-man-pages.py:
```diff
@@ -0,0 +1,73 @@
+    build_man_pages(args.version, args.output_dir)
+
+
+def build_man_pages(version: str, output_dir: pathlib.Path) -> None:
```
Add a default to `output_dir` so that an external caller that sets a CWD, wouldn't have to also calculate this path.

webknjaz commented on packaging/pep517_backend/_backend.py:
```diff
@@ -8,9 +8,8 @@
-from importlib.metadata import import_module
-from io import StringIO
+from contextlib import contextmanager
+from importlib import import_module
```
I'd keep refactoring this separate so it doesn't create confusion about being mandatory for the standalone manpage builder to work..

webknjaz commented on hacking/build-man-pages.py:
```diff
@@ -0,0 +1,73 @@
+
+    subprocess.check_call(generate_man_cmd)
+
+    return list(output_dir.glob('*.1.rst.in'))
```
No need to turn this into a list since it's being iterated over exactly once.

webknjaz commented on hacking/build-man-pages.py:
```diff
@@ -0,0 +1,73 @@
+        CHECKOUT_DIR / 'hacking/build-ansible.py',
+        'generate-man',
+        '--template-file', CHECKOUT_DIR / 'docs/templates/man.j2',
+        '--output-dir', output_dir,
```
I originally used the `--option=value` syntax specifically to avoid those problems that some formatters tend to put such pairs on separate lines...

webknjaz reviewed (commented):
I was going to respond that I'm partially okay with this but as I was writing the review, I realized that I'm torn. I think that, at least for now, we need to split the effort into (1) refactoring and (2) adding a CLI.
Sustainability of (2) is still questionable. And (1) could probably be done better.

We'll need to define the boundaries between what we can commit to supporting (the PEP 517-driven flow) vs. what happens to work for now and any users must accept that it may stop working anytime with conscious understanding that they are playing with fire.

Anything relating to (1) **MUST** be tested automatically.

For now, with the structure that I have in mind, this could be a one-liner for the Fedora downstream like `PYTHONPATH=packaging/ python -c 'from pep517_backend.api.manpage_builder import build_manpages; build_manpages()'`.

I think I might suggest a PR with the refactoring soon. Anyway, I've kept some of the in-diff comments I wrote before composing this review summary, but they are not entirely complete.

gotmax23 commented:
Is there any update here? What still needs to be fixed?

gotmax23 commented:
@webknjaz Hmm, it seems your response got lost in my email trove. Thanks for the update! In the interim, would you at least accept a PR to split

https://github.com/ansible/ansible/blob/68e270d4cc2579e4659ed53aecbc5a3358b85985/packaging/pep517_backend/_backend.py#L125-L133

into a separate function in the same file so I can call it with `python -c` in the specfile?

webknjaz commented:
No, I don't think so. I already have a series of commits with most of the changes lined up in my local branch. It does separate it out and even more but making parallel changes would create conflicts.
I'm at PyCon US now but I'll try to make time for the PR. In the meantime, since you clone from Git anyway, you could just reset the corresponding files to older versions and run those, I suppose.

mattclay commented:
With the split of the `docs` directory into a separate repository (and the associated refactoring) this PR doesn't make sense as-is (and has merge conflicts), so I'm closing it. I still think a solution for https://github.com/ansible/ansible/issues/80368 would be good, but that's likely to come out of an overhauled version of https://github.com/ansible/ansible/pull/80592 or a fresh PR.
